packages/compiler/core/scanner.ts

```diff
@@ -469,72 +520,157 @@ export function createScanner(
+          if (atConflictMarker()) return scanConflictMarker();
+          return lookAhead(1) === CharCode.Bar ? next(Token.BarBar, 2) : next(Token.Bar);
 
         case CharCode.DoubleQuote:
           return lookAhead(1) === CharCode.DoubleQuote && lookAhead(2) === CharCode.DoubleQuote
             ? scanTripleQuotedString()
             : scanString();
+
         case CharCode.Exclamation:
           return lookAhead(1) === CharCode.Equals
             ? next(Token.ExclamationEquals, 2)
             : next(Token.Exclamation);
+
         default:
           if (isLowercaseAsciiLetter(ch)) {
             return scanIdentifierOrKeyword();
           }
-
           if (isAsciiIdentifierStart(ch)) {
             return scanIdentifier();
           }
-
           if (ch <= CharCode.MaxAscii) {
             return scanInvalidCharacter();
           }
-
           return scanNonAsciiToken();
       }
     }
 
     return (token = Token.EndOfFile);
   }
 
-  function next(t: Token, count = 1) {
+  function scanDoc(): DocToken {
+    tokenPosition = position;
+    tokenFlags = TokenFlags.None;
+
+    if (!eof()) {
+      const ch = input.charCodeAt(position);
+      switch (ch) {
+        case CharCode.CarriageReturn:
+          if (lookAhead(1) === CharCode.LineFeed) {
+            position++;
+          }
+        // fallthrough
+        case CharCode.LineFeed:
+          return next(Token.NewLine);
+
+        case CharCode.Space:
+        case CharCode.Tab:
+        case CharCode.VerticalTab:
+        case CharCode.FormFeed:
+          return scanWhitespace();
+
+        case CharCode.CloseBrace:
+          return next(Token.CloseBrace);
+
+        case CharCode.At:
+          return next(Token.At);
+
+        case CharCode.Asterisk:
+          return next(Token.Star);
+
+        case CharCode.Backtick:
+          return lookAhead(1) === CharCode.Backtick && lookAhead(2) === CharCode.Backtick
+            ? next(Token.DocCodeFenceDelimiter, 3)
+            : scanDocCodeSpan();
+
+        case CharCode.LessThan:
+        case CharCode.GreaterThan:
+        case CharCode.Equals:
+        case CharCode.Bar:
+          if (atConflictMarker()) return scanConflictMarker();
+          return next(Token.DocText);
+      }
+
+      if (isAsciiIdentifierStart(ch)) {
+        return scanIdentifier();
+      }
+
+      if (ch <= CharCode.MaxAscii) {
+        return next(Token.DocText);
+      }
+
+      const cp = input.codePointAt(position)!;
+      if (isIdentifierStart(cp)) {
+        return scanNonAsciiIdentifier(cp);
+      }
+
+      return scanUnknown(Token.DocText);
+    }
+
+    return (token = Token.EndOfFile);
+  }
+
+  function scanRange<T>(range: TextRange, callback: () => T): T {
+    const savedPosition = position;
+    const savedEndPosition = endPosition;
+    const savedToken = token;
+    const savedTokenPosition = tokenPosition;
+    const savedTokenFlags = tokenFlags;
+    const savedMode = mode;
+
+    let result: T;
+    try {
```

Comment:
I'm going to remove this try/finally. TS doesn't do this and no other scan method tries to be transactional vs. exceptions. If the scanner throws, it should be considered done.